flow/stopwatch_dl.cc

```diff
@@ -0,0 +1,123 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/stopwatch_dl.h"
+#include "display_list/dl_blend_mode.h"
+#include "display_list/dl_canvas.h"
+#include "display_list/dl_paint.h"
+#include "include/core/SkPath.h"
+#include "include/core/SkRect.h"
+
+namespace flutter {
+
+static const size_t kMaxSamples = 120;
+static const size_t kMaxFrameMarkers = 8;
+
+void DlStopwatchVisualizer::Visualize(DlCanvas* canvas,
+                                      const SkRect& rect) const {
+  DlPaint paint;
+
+  // Establish the graph position.
+  auto const x = rect.x();
+  auto const y = rect.y();
+  auto const width = rect.width();
+  auto const height = rect.height();
+
+  // Scale the graph to show time frames up to those that are 3x the frame time.
+  auto const one_frame_ms = stopwatch_.GetFrameBudget().count();
+  auto const max_interval = one_frame_ms * 3.0;
+  auto const max_unit_interval = UnitFrameInterval(max_interval);
+  auto const sample_unit_width = (1.0 / kMaxSamples);
+
+  // Erase all pixels.
+  {
+    paint.setColor(0x99FFFFFF);
+    canvas->DrawRect(rect, paint);
+  }
+
+  // Prepare a path for the data; we start at the height of the last point so
+  // it looks like we wrap around.
+  {
+    SkPath path;
+    path.setIsVolatile(true);
+    path.moveTo(x, rect.bottom());
+    path.lineTo(
+        x,
+        y + height * (1.0 - UnitHeight(stopwatch_.GetLap(0).ToMillisecondsF(),
+                                       max_unit_interval)));
+
+    double unit_x;
+    double next_x = 0.0;
+    for (auto i = size_t(0); i < stopwatch_.GetLapsCount(); i++) {
+      unit_x = next_x;
+      next_x = static_cast<double>(i + 1) / kMaxSamples;
+
+      auto const sample_y =
+          y + height * (1.0 - UnitHeight(stopwatch_.GetLap(i).ToMillisecondsF(),
+                                         max_unit_interval));
+      path.lineTo(x + width * unit_x, sample_y);
+      path.lineTo(x + width * next_x, sample_y);
+    }
+
+    path.lineTo(
+        rect.right(),
+        y + height *
+                (1.0 - UnitHeight(
+                           stopwatch_.GetLap(kMaxSamples - 1).ToMillisecondsF(),
+                           max_unit_interval)));
+    path.lineTo(rect.right(), rect.bottom());
+    path.close();
+
+    paint.setColor(0xAA0000FF);
+    canvas->DrawPath(path, paint);
+  }
+
+  // Draw horizontal frame markers.
+  {
+    paint.setStrokeWidth(0);
```

Comment:
Same here, use skinny rectangles all packed into the same vertices instead of a path/stroke.

Reply:
Ditto!